proposals/1406_multiple-idps/README.md

```diff
@@ -0,0 +1,493 @@
+  if usernames in that IDP are not allowed to contain a "/" character.
+
+Transformation expressions on the FederationDomain can be easily used to avoid identity collisions as desired.
+For example, the CEL expressions to prefix every username and group name are `"my-prefix:" + username` and
+`groups.map(g, "my-prefix:" + g)`.
+
+#### Upgrades
+
+Any upgrades into a new version of Pinniped which allows multiple IDPs will have a similar configuration. There will
+be a FederationDomain with no IDPs listed on the FederationDomain (since this was not previously allowed), and there
+will be only a single IDP CRD created in the namespace. Any other number of IDP CRDs previously resulted in an
+unusable Pinniped installation.
+
+During an upgrade, an existing installation of the Supervisor would already have a FederationDomain CR defined without
+an "identityProviders" section. To enable smooth upgrades, the "identityProviders" section would be optional.
+
+- The Supervisor code already correctly handles the case when there are no identity provider CRs defined. No users can
+  log in using that FederationDomain.
+- To handle the case where there is exactly one identity provider CR defined, the controller could load that CR for use
+  in the FederationDomain. The "displayName" of the identity provider would be automatically configured to be the same
+  name as the CR. This allows old configurations to continue working after upgrade.
+- When there are multiple identity provider CRs defined, the controller can fail to load the FederationDomain and update
+  its status to include an error saying that using a FederationDomain when multiple identity provider CRs are created
+  requires using the "identityProviders" field on the FederationDomain. This handles the case where the
+  user adds multiple identity provider CRs after upgrading, but forgets to add the "identityProviders" field to the
+  FederationDomain.
+
+If an admin adds "identityProviders" to a pre-existing FederationDomain and changes the "displayName" of a pre-existing
+identity provider, then:
+1. Pre-existing user sessions would fail to refresh, causing those users to need to interactively log in again, since
+   the identity provider names and types are already stored in user sessions for use during refreshes. This code already
+   has sufficient protections to ensure that we can never accidentally use a different identity provider during refresh
+   compared to which was used during initial login, even if there is an accidental name collision (via UID comparisons).
+2. Pre-existing kubeconfigs would now refer to the wrong identity provider name, and would need to be regenerated.
+
+If an admin wants to add a pre-existing identityProvider to a pre-existing FederationDomain without interrupting
+pre-existing sessions or needing to generate new kubeconfigs, they could take care to make the "displayName" of
+the identity provider exactly match the name of the identity provider CR.
```

Comment:
I think we should state in this proposal that admins will be able to generate all possible Kubeconfigs for all possible IDPs for a given cluster via a single command (`pinniped get kubeconfig --all-idps` or something like that).

Reply:
I think this is also potentially a good idea. Can we make this a separate feature to consider after the first version?